Mudlet/Mudlet
# Add: Prompt for secure connection choice

#### Brief overview of PR additions

To encourage enhanced data transfer protection and privacy, respond to detection of the `TLS` (or `SSL` legacy) key of MSSP ([Mud Server Status Protocol](https://wiki.mudlet.org/w/Manual:Technical_Manual#MSSP)) and prompt a user not on a TLS connection with a choice to reconnect with the advertised TLS port from MSSP.

<img width="924" alt="Screen Shot 2022-11-05 at 7 03 05 PM" src="https://user-images.githubusercontent.com/3058178/200146744-9aa3305d-99b2-4e4b-ab48-99779d7e0f76.png">

If the user selects `Yes`, automatically update the port with the `TLS` value gathered from MSSP, check-mark the `Secure` checkbox on the connection dialog, then reconnect.  If the user selects `No`, automatically update a profile preference so they are not asked again for the current profile, then reconnect.  This preference may be controlled on the Settings->Connection menu.  This preference is enabled by default.

<img width="990" alt="Screen Shot 2022-11-05 at 7 37 14 PM" src="https://user-images.githubusercontent.com/3058178/200147085-89b0066f-e91b-4b89-8f3f-450c7b407bda.png">

- MSSP `TLS` key values of `-1` or `0` are indicators the game does not offer a TLS port.
- If an `TLS` (or legacy `SSL`) key/value pair does not exist, nothing happens.

#### Motivation for adding to Mudlet

Data protection and privacy.

#### Other info (issues closed, discussion etc)

Closes Issue #5075 [Make use of MSSP's secure connection flag](https://github.com/Mudlet/Mudlet/issues/5075)

##### Release Notes

To encourage enhanced data transfer protection and privacy, users will be prompted for a choice to switch to an encrypted port when this is advertised by their game via the Mud Server Status Protocol (MSSP).

- Many more users will shift over easily to secure ports, which are little known to others than their game runners.
- Increasing awareness of the `TLS` key of MSSP among game runners will be needed.

[Wiki Updated](https://wiki.mudlet.org/w/Manual:Supported_Protocols#Secure_connection_prompt)

##### How to Test

1. Connect to host: `stickmud.com` port: `7680` with the `Secure` checkbox unchecked.  If you use the default game button in Mudlet, be sure to make the appropriate updates specified previously.
2. Receive the prompt, select `Yes`, then verify you receive the secure connection confirmations.
3. Close Mudlet, open Mudlet, connect to StickMUD, and note the port is changed to `7670` and the `Secure` checkbox is checked.  There should be no prompt and you will continue to the logon screen of StickMUD.
4. Close Mudlet, open Mudlet, change the port back to `7680` and uncheck the `Secure` checkbox, then connect to StickMUD.
5. Receive the prompt, select `No`, then connect to StickMUD without receiving the secure connection confirmations.
6. Open up *Settings->Connection* and note that `Allow secure connection reminder` is unchecked.
7. Close Mudlet, open Mudlet, then connect to StickMUD without receiving the prompt.
8. Open up *Settings->Connection* and checkmark `Allow secure connection reminder` and hit the `Save` button.
9. Close Mudlet, open Mudlet, change the port back to `7680`, then connect to StickMUD.
10. Receive the prompt.

##### Design Decisions

Although it would be preferred to not reconnect when the user selects `No`, the buffer would not continue writing without hitting the return key afterward.  Since this scenario only happens once per profile, the approach to reconnect was selected.

This PR could be followed up with an enhancement to test the TLS port before activating this scenario, but the control offered by the profile preference should offer a good work-around to solve any outlier edge cases and that may never be needed.

##### Miscellaneous

At the time of submitting this PR, there was an issue with timestamps appearing upon the second profile load which is reported in Issue #6408.
